Fix unit test on Python 3.8

bdrung opened the pull request:
Salt fails on Python 3.8:

```
======================================================================
ERROR: unit.grains.test_core (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: unit.grains.test_core
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.8/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "tests/unit/grains/test_core.py", line 37, in <module>
    import salt.grains.core as core
  File "salt/grains/core.py", line 40, in <module>
    from platform import _supported_dists
ImportError: cannot import name '_supported_dists' from 'platform' (/usr/lib/python3.8/platform.py)
```

So only try to import `_supported_dists` from `platform` for Python <=
3.7. Otherwise rely on the external `distro` module to  not need any
special handling.

Addresses parts of #55835

bdrung committed:
Support distro.linux_distribution

Salt fails on Python 3.8:

```
======================================================================
ERROR: unit.grains.test_core (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: unit.grains.test_core
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.8/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "tests/unit/grains/test_core.py", line 37, in <module>
    import salt.grains.core as core
  File "salt/grains/core.py", line 40, in <module>
    from platform import _supported_dists
ImportError: cannot import name '_supported_dists' from 'platform' (/usr/lib/python3.8/platform.py)
```

So only try to import `_supported_dists` from `platform` for Python <=
3.7. Otherwise rely on the external `distro` module to  not need any
special handling.

Addresses parts of #55835
Signed-off-by: Benjamin Drung <benjamin.drung@cloud.ionos.com>

bdrung committed:
Fix RuntimeError: dictionary keys changed during iteration

The following unit tests fail on Python 3.8:

```
======================================================================
ERROR: test_state_config (unit.renderers.test_stateconf.StateConfigRendererTestCase)
[CPU:0.0%|MEM:56.6%]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/tests/unit/renderers/test_stateconf.py", line 74, in test_state_config
    result = self._render_sls('''
  File "/<<PKGBUILDDIR>>/tests/unit/renderers/test_stateconf.py", line 66, in _render_sls
    return self._renderers['stateconf'](
  File "/<<PKGBUILDDIR>>/salt/renderers/stateconf.py", line 227, in render
    for k in six.iterkeys(tmplctx):  # iterate over a copy of keys
RuntimeError: dictionary keys changed during iteration

======================================================================
ERROR: test_apply_cloud_providers_config_extend (unit.test_config.ConfigTestCase)
[CPU:0.0%|MEM:56.6%]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/tests/unit/test_config.py", line 1243, in test_apply_cloud_providers_config_extend
    salt.config.apply_cloud_providers_config(
  File "/<<PKGBUILDDIR>>/salt/config/__init__.py", line 3196, in apply_cloud_providers_config
    for driver, details in six.iteritems(entries):
RuntimeError: dictionary keys changed during iteration

======================================================================
ERROR: test_apply_cloud_providers_config_extend_multiple (unit.test_config.ConfigTestCase)
[CPU:0.0%|MEM:56.6%]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/tests/unit/test_config.py", line 1334, in test_apply_cloud_providers_config_extend_multiple
    self.assertEqual(ret, salt.config.apply_cloud_providers_config(overrides, defaults=DEFAULT))
  File "/<<PKGBUILDDIR>>/salt/config/__init__.py", line 3196, in apply_cloud_providers_config
    for driver, details in six.iteritems(entries):
RuntimeError: dictionary keys changed during iteration

======================================================================
```

Replace the affected for loop of the first case by a dictionary
comprehension to construct the modified dictionary. For the remaining
cases, switch from `iteritems` to `iterkeys`, since the dictionary
values will be modified.

Signed-off-by: Benjamin Drung <benjamin.drung@cloud.ionos.com>

bdrung committed:
Fix unit.modules.test_virt.VirtTestCase.test_update

The test case `unit.modules.test_virt.VirtTestCase.test_update` fails on
Ubuntu 20.04 with Python 3.8:

```
======================================================================
FAIL: test_update (unit.modules.test_virt.VirtTestCase)
[CPU:0.0%|MEM:17.0%]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/unit/modules/test_virt.py", line 1313, in test_update
    self.assertEqual({
AssertionError: {'definition': False, 'disk': {'attached': [], 'detached[49 chars] []}} != {'definition': True, 'disk': {'attached': [], 'detached'[74 chars]True}
+ {'cpu': True,
- {'definition': False,
? ^              ^^^^

+  'definition': True,
? ^              ^^^

   'disk': {'attached': [], 'detached': []},
-  'interface': {'attached': [], 'detached': []}}
?                                               ^

+  'interface': {'attached': [], 'detached': []},
?                                               ^

+  'mem': True}

======================================================================
```

`virt.update` falsely detects a change to the graphics setting in the
"no diff case". The old XML part specifies:

```xml
<graphics type="spice" port="5900" autoport="yes" listen="127.0.0.1">
    <listen type="address" address="127.0.0.1" />
</graphics>
```

The newly generated XML misses the port setting:

```xml
<graphics type="spice" listen="127.0.0.1" autoport="yes">
    <listen type="address" address="127.0.0.1" />
</graphics>
```

So adjust the initial XML file of the template, because setting the port
to the default 5900, will change `autoport` to `no`.

Signed-off-by: Benjamin Drung <benjamin.drung@cloud.ionos.com>

bdrung commented:
I pushed more commits to fix all Python 3.8 related unit tests failures.

s0undt3ch committed:
Update PyTestSalt requirement(because we now bundle tornado)

s0undt3ch committed:
Run the full test suite on Arch under Py3

s0undt3ch commented:
FYI, I also did some work on this [here](https://github.com/saltstack/salt/tree/features/py38-compat).
I'm not devoting much time to it right now, so feel free to rebase your work on top of that branch, it would be **much** appreciated.

bdrung commented:
I picked the commits from your branch that I haven't addressed yet.

bdrung committed:
Fix deprecation warnings for imports from collections

DeprecationWarning: Using or importing the ABCs from `collections`
instead of from `collections.abc` is deprecated since Python 3.3, and in
3.9 it will stop working.

Therefore try to import the abstract base classes from `collections.abc`
before falling back to `collections`.

Signed-off-by: Benjamin Drung <benjamin.drung@cloud.ionos.com>

cmcmarrow reviewed (changes requested):
@bdrung thanks for doing this! You’re really close. Just need to get the Windows and Mac test working. Once you have that I think this PR will be good to merge.

bdrung commented:
The failing Windows and Mac tests are identical with the failings tests of the master branch. So this merge request does not introduce any regression.

Ch3LL commented:
those tests should be resolved on the head of master so i updated the branch.

waynew commented on .pre-commit-config.yaml:
```diff
@@ -306,6 +306,203 @@ repos:
+          - --include=requirements/pytest.txt
+          - --passthrough-line-from-input=^pyobjc(.*)$
+
+#      - id: pip-tools-compile
```
@s0undt3ch can we add a comment to this part explaining when it needs to be un-commented? I'm okay with pre-staging in this file, but we need something to remind us that it needs to be re-added.

Either that, or just remove it from this PR and open up a separate one that adds it in.

s0undt3ch replied:
Done

waynew commented on .pre-commit-config.yaml:
```diff
@@ -306,6 +306,203 @@ repos:
+          - --platform=darwin
+          - --out-prefix=darwin
+
+#      - id: pip-tools-compile
```
@s0undt3ch same here

waynew commented on .pre-commit-config.yaml:
```diff
@@ -306,6 +306,203 @@ repos:
+          - --include=requirements/pytest.txt
+          - --passthrough-line-from-input=^pyobjc(.*)$
+
+#      - id: pip-tools-compile
```
@s0undt3ch and here

waynew commented on .pre-commit-config.yaml:
```diff
@@ -306,6 +306,203 @@ repos:
+          - --platform=darwin
+          - --out-prefix=darwin
+
+#      - id: pip-tools-compile
```
@s0undt3ch and here

s0undt3ch commented:
FYI, I pushed a commit that fixes requirements.

bdrung commented:
Thanks for adding the fix. I rebased the branch on master and fixed two commits that needed blackening.

s0undt3ch commented:
Fixed black and pre-commit is passing now

waynew reviewed (dismissed):
👍

terminalmage commented on salt/grains/core.py:
```diff
@@ -59,29 +59,34 @@
-            return _deprecated_linux_distribution(**kwargs)
+            return _deprecated_linux_distribution(
+                supported_dists=_supported_dists, **kwargs
+            )
```
:+1:

bdrung commented:
I rebased this patch set on master to resolve the merge commits. This rebase dropped some commits which are already part of the master branch.

s0undt3ch commented:
It has some lint issues that need fixing now

bdrung commented:
Fixed the lint issues.

waynew commented:
applies to #55835

OrangeDog commented:
Why did this bother with the try/except on `collections.abc`, when the old location has been deprecated since 3.3, and the minimum supported version for Sodium is 3.5?

bdrung commented:
Good point. Please open a merge request to drop falling back to the old location.
